### DISCUSSION TITLE
Problem related to Grafana metrics  UI

### DISCUSSION THREAD

**Discussion by hakegard** [srsran/srsRAN_Project#418] [General Help] [answered]
Hello,

I try to run the Grafana metric UI following the guide in the documentation. When enabling reporting metrics in JSON format in the gnb configuration file by inserting the four lines from the guide, I get the error: "INI was not able to parse metrics.enable_json_metrics" when starting the gnb. Leaving the enable_json_metrics parameter empty passes, while setting it to true or false leads to the same error. Does anyone have an explanation or solution?

**Comment by brendan-mcauliffe** [comment]
Hey @hakegard, 

Are you using the latest version of the code? This was added in a recent update.

**Reply by hakegard** [reply]
Hi brendan, 

Thanks, you were correct. I thought I had the last version, but I did not. Now it works perfectly!

**Reply by hakegard** [reply]
There is however one problem. When Installing grafana, internet connectivity is lost. Using wireshark, it appears that when pinging an external IP-address, the ICMP packets don't reach the WAN interface. Completely removing docker solves the problem. Maybe it is something obvious I'm missing, but I'm not able to see it.

I'm setting up the tun interface and  WAN connectivity as follows:
$sudo ip tuntap add name ogstun mode tun
$sudo ip addr add 10.45.0.1/16 dev ogstun
$sudo ip addr add 2001:db8:cafe::1/48 dev ogstun
$sudo ip link set ogstun up

$sudo sysctl -w net.ipv4.ip_forward=1
$sudo sysctl -w net.ipv6.conf.all.forwarding=1
$sudo iptables -t nat -A POSTROUTING -s 10.45.0.0/16 ! -o ogstun -j MASQUERADE
$sudo ip6tables -t nat -A POSTROUTING -s 2001:db8:cafe::/48 ! -o ogstun -j MASQUERADE

The routing tables without docker installed looks like this:
```
 $route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         _gateway        0.0.0.0         UG    100    0        0 enp67s0f0
10.45.0.0       0.0.0.0         255.255.0.0     U     0      0        0 ogstun
10.219.45.0     0.0.0.0         255.255.255.192 U     100    0        0 enp67s0f0
link-local      0.0.0.0         255.255.0.0     U     1000   0        0 enp67s0f0
```
With grafana/docker installed, it looks like this:
```
$ route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         _gateway        0.0.0.0         UG    100    0        0 enp67s0f0
10.45.0.0       0.0.0.0         255.255.0.0     U     0      0        0 ogstun
10.53.1.0       0.0.0.0         255.255.255.0   U     0      0        0 br-96a56d86f651
10.219.45.0     0.0.0.0         255.255.255.192 U     100    0        0 enp67s0f0
link-local      0.0.0.0         255.255.0.0     U     1000   0        0 enp67s0f0
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
172.19.1.0      0.0.0.0         255.255.255.0   U     0      0        0 br-88a994db56f9
```

**Reply by asaezper** [reply, marked as answer]
hey!

In the docker-compose, there are two docker networks defined:
```yml
networks:
  ran:
    ipam:
      driver: default
      config:
        - subnet: 10.53.1.0/24
  metrics:
    ipam:
      driver: default
      config:
        - subnet: 172.19.1.0/24
```

After a docker compose up, docker will create those two networks. That's the reason why you see entries for them in your routing table. I don't see how they can overlap with your previously existing ones, but looks like that's what happening. 

I'd say:
- Check first that everything works after docker is installed and before creating those networks.
- Change subnet ranges in docker-compose.yml to different ones, picking an empty range for you. They should be in private range.
- If you have issues, you don't need to remove docker, it should be enough to remove the network with `docker network rm <name>` Also, you could use `docker system prune -a --volumes -f` which removes containers, images and networks.

**Reply by pgawlowicz** [reply]
@hakegard any update on this issue?

**Reply by hakegard** [reply]
@pgawlowicz  No updates. Docker (and grafana) can be installed and works without breaking internet connectivity until the computer is rebooted. After reboot, interconnectivity is broken as described above. So a workaround is to uninstall grafana and docker before rebooting and reinstalling after reboot. Not very convenient, but works if the machine is not rebooted too often...

**Reply by pgawlowicz** [reply]
How do you start grafana? It should be like this:
```
sudo docker compose up grafana
```
Also, do you use `10.53.1.0/24` or `172.19.1.0/24` addresses in your network?

**Reply by hakegard** [reply]
No, I don't. And as I tried to explain below, the problem occurs wtihout grafana, or any other containers, installed. So it is a problem related to docker and not grafana. Installing docker and rebooting routes the packets from the UE to the lo interface instead of to the ethernet interface.

**Reply by pgawlowicz** [reply]
does it help if you keep docker installed, but run `docker system prune -a`. Note that this command removes containers, images and networks.

**Reply by hakegard** [reply]
No , it does unfortunately not help. 
- Without docker installed, routing to the ethernet interface works and the UE has internet access.
- Installing docker with 'sudo apt install docker-ce' (i.e. without any containers, images and networks), routing to the ethernet interface works and the UE has internet access.
- Rebouting after docker is installed: UE packets are routed to the lo interface, and internet access is lost.

**Reply by pgawlowicz** [reply]
ok, then not sure what is the problem. Usually, docker does not interfere with the networking configuration. Could you print the routing on the UE node?

**Reply by hakegard** [reply]
There are a few difference in the routing table of the UE for the two cases (docker installed (1) and not installed (2) at the base station) that I have not noticed before:
```
(1)$ route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         10.45.0.5       0.0.0.0         UG    21050  0        0 wwx02219ea4feeb
10.45.0.0       0.0.0.0         255.255.255.248 U     1050   0        0 wwx02219ea4feeb
link-local      0.0.0.0         255.255.0.0     U     1000   0        0 wwx02219ea4feeb
```
```
(2)$ route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         _gateway        0.0.0.0         UG    1050   0        0 wwx02219ea4feeb
10.45.0.0       0.0.0.0         255.255.255.252 U     1050   0        0 wwx02219ea4feeb
link-local      0.0.0.0         255.255.0.0     U     1000   0        0 wwx02219ea4feeb
```

**Reply by pgawlowicz** [reply]
hmm, do you use COTS UE or srsUE?

**Reply by hakegard** [reply]
We use COTS UEs. The routing is from a laptop with a Simcom modem connecting to the base station. We also use OnePlus phones with the same result.

**Reply by pgawlowicz** [reply]
could you share your gnb.conf file? Do you use docker only for Grafana or also for Open5gs?

**Reply by hakegard** [reply]
I'm not using docker for open5gs. The conf file is attached.

[gnb.yaml.txt](https://github.com/srsran/srsRAN_Project/files/14261313/gnb.yaml.txt)

**Reply by pgawlowicz** [reply]
hmm, I am not sure what is going on. There should be no routing change at UE depending on whether docker is installed or not. 
Could you try replace the manually installed Open5gs with the dockerized version? 
see details here: 
https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html#open5gs-core
and you need to change the amf address in the gnb config. 

also check here if case of issues:
https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html#nd-open5gs-instance-installed-manually

**Comment by hakegard** [comment]
Hi @asaezper,

Thanks for the response. Unfortunately, the problem occurs also without any containers or images. The routing table is as expected:
```
$ route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         _gateway        0.0.0.0         UG    100    0        0 enp67s0f0
10.45.0.0       0.0.0.0         255.255.0.0     U     0      0        0 ogstun
10.219.45.0     0.0.0.0         255.255.255.192 U     100    0        0 enp67s0f0
link-local      0.0.0.0         255.255.0.0     U     1000   0        0 enp67s0f0
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
```
I was wondering if it the problem was related to postrouting, as docker0 appers in iptables: 
```
$ sudo iptables -t nat -v -L POSTROUTING -n --line-number
[sudo] password for sep5g: 
Chain POSTROUTING (policy ACCEPT 0 packets, 0 bytes)
num   pkts bytes target     prot opt in     out     source               destination         
1        0     0 MASQUERADE  all  --  *      !docker0  172.17.0.0/16        0.0.0.0/0           
2        0     0 MASQUERADE  all  --  *      !ogstun  10.45.0.0/16         0.0.0.0/0  
```
However, when pinging external IP-addresses from the UE with docker installed, I see that the ICMP (ping request) packets are routed to the lo interface using wireshark, while they are correctly routed to the ethernet interface without docker installed. Is it possible that docker intervene with the routing, or that the srsran code detects that docker is installed and changes the routing?

**Comment by VikrantDigiCat** [comment]
- I followed the DPDK install guide (https://docs.srsran.com/projects/project/en/latest/tutorials/source/dpdk/source/index.html).
- Have set the metrics section in my gnb config file as below:
<img width="928" alt="image" src="https://github.com/user-attachments/assets/7d90d898-939a-428a-855b-d0eea16f313c" />

- But getting error as below:
<img width="339" alt="image" src="https://github.com/user-attachments/assets/926feeeb-9f86-491a-b786-187fc69f97fd" />

Can anyone suggest what's wrong here?

**Reply by brendan-mcauliffe** [reply]
Hi @VikrantDigiCat, we have updated the configuration reference file [here](https://docs.srsran.com/projects/project/en/latest/user_manuals/source/config_ref.html) to align it with the new configuration parameters released with the latest version of srsRAN. This should fix the above issue. Please take a look and let me know.
